includes/service-providers/class-newspack-newsletters-service-provider.php

```diff
@@ -0,0 +1,278 @@
+<?php
+/**
+ * Service Provider: Mailchimp Implementation
+ *
+ * @package Newspack
+ */
+
+defined( 'ABSPATH' ) || exit;
+
+/**
+ * Main Newspack Newsletters Class.
+ */
+abstract class Newspack_Newsletters_Service_Provider {
+
+	/**
+	 * Class constants.
+	 */
+	const BASE_NAMESPACE = 'newspack-newsletters/v1/';
+
+	/**
+	 * Class properties.
+	 */
+
+	/**
+	 * Name of the service.
+	 *
+	 * @var string
+	 */
+	public $service;
+
+	/**
+	 * Instances of descendant service provider classes.
+	 *
+	 * @var array
+	 */
+	protected static $instances = [];
+
+	/**
+	 * Abstract methods.
+	 */
+
+	/**
+	 * Set list for a campaign.
+	 *
+	 * @param string $post_id Campaign Id.
+	 * @param string $list_id ID of the list.
+	 * @return object|WP_Error API API Response or error.
+	 */
+	abstract public function list( $post_id, $list_id );
+
+	/**
+	 * Retrieve a campaign.
+	 *
+	 * @param integer $post_id Numeric ID of the Newsletter post.
+	 * @return object|WP_Error API Response or error.
+	 */
+	abstract public function retrieve( $post_id );
+
+	/**
+	 * Update ESP campaign after post save.
+	 *
+	 * @param string  $post_id Numeric ID of the campaign.
+	 * @param WP_Post $post The complete post object.
+	 * @param boolean $update Whether this is an existing post being updated or not.
+	 */
+	abstract public function save( $post_id, $post, $update );
+
+	/**
+	 * Send a campaign.
+	 *
+	 * @param integer $post_id Post ID to send.
+	 * @param WP_POST $post Post to send.
+	 */
+	abstract public function send( $post_id, $post );
+
+	/**
+	 * Set sender data.
+	 *
+	 * @param string $post_id Numeric ID of the campaign.
+	 * @param string $from_name Sender name.
+	 * @param string $reply_to Reply to email address.
+	 * @return object|WP_Error API Response or error.
+	 */
+	abstract public function sender( $post_id, $from_name, $reply_to );
+
+	/**
+	 * Synchronize post with corresponding ESP campaign.
+	 *
+	 * @param WP_POST $post Post to synchronize.
+	 * @return object|null API Response or error.
```

Comment:
Shouldn't it be
```suggestion
	 * @return object|WP_Error API Response or error.
```
?

Reply:
I used `null` because of this possible outcome: https://github.com/Automattic/newspack-newsletters/blob/4e9a31ce9a3baf74e3f07fdc773483207267d175/includes/service-providers/mailchimp/class-newspack-newsletters-mailchimp.php#L340. Because of this approach to storing errors for future notification perhaps https://github.com/Automattic/newspack-newsletters/blob/4e9a31ce9a3baf74e3f07fdc773483207267d175/includes/service-providers/mailchimp/class-newspack-newsletters-mailchimp.php#L287-L292 should be changed to persist the error and return `null`?

Reply:
Ah, that's true that there's a persisted error there, not returned one. I think it's ok. My comment was prompted more by the `or error` wording :)